File: import_pcms_data.flow/finalize_lineage.inline_script.ts
/**
 * Finalize Lineage - Aggregates results from all import steps
 * 
 * This step runs last and:
 * 1. Aggregates errors from all previous steps
 * 2. Counts total records processed
 * 3. Returns final summary
 */

// ─────────────────────────────────────────────────────────────────────────────
// Types
// ─────────────────────────────────────────────────────────────────────────────

interface TableResult {
  table: string;
  attempted: number;
  success: boolean;
  error?: string;
}

interface StepSummary {
  dry_run?: boolean;
  tables?: TableResult[];
  errors?: string[];
}

// ─────────────────────────────────────────────────────────────────────────────
// Main
// ─────────────────────────────────────────────────────────────────────────────

export async function main(
  dry_run = false,
  summaries: StepSummary[] = []
) {
  const startedAt = new Date().toISOString();

  // Aggregate errors and record counts from all step summaries
  const allErrors: string[] = [];
  let recordCount = 0;

  for (const s of summaries) {
    // Collect errors array from each step
    for (const err of s?.errors ?? []) {
      allErrors.push(err);
    }

    // Collect table-level errors and count records
    for (const t of s?.tables ?? []) {
      recordCount += Number(t.attempted ?? 0);
      if (!t.success) {
        allErrors.push(`${t.table}: ${t.error ?? "unknown error"}`);
      }
    }
  }

  const status = allErrors.length > 0 ? "FAILED" : "SUCCESS";

  return {
    dry_run,
    started_at: startedAt,
    finished_at: new Date().toISOString(),
    tables: [],
    errors: allErrors,
    status,
    record_count: recordCount,
    error_count: allErrors.length,
  };
}

File: import_pcms_data.flow/finalize_lineage.inline_script.py
"""
Finalize Lineage - Aggregates results from all import steps

This step runs last and:
1. Aggregates errors from all previous steps
2. Counts total records processed
3. Returns final summary

No database operations - just summarizes results.
"""
from datetime import datetime

# ─────────────────────────────────────────────────────────────────────────────
# Main
# ─────────────────────────────────────────────────────────────────────────────


def main(dry_run: bool = False, summaries: list[dict] = None):
    """
    Aggregate results from all import steps.

    Args:
        dry_run: Whether this was a dry run
        summaries: List of step summaries from previous steps
    """
    started_at = datetime.now().isoformat()
    summaries = summaries or []

    # Aggregate errors and record counts from all step summaries
    all_errors: list[str] = []
    record_count = 0

    for s in summaries:
        # Collect errors array from each step
        for err in s.get("errors") or []:
            all_errors.append(err)

        # Collect table-level errors and count records
        for t in s.get("tables") or []:
            record_count += int(t.get("attempted") or 0)
            if not t.get("success"):
                all_errors.append(f"{t.get('table')}: {t.get('error', 'unknown error')}")

    status = "FAILED" if all_errors else "SUCCESS"

    return {
        "dry_run": dry_run,
        "started_at": started_at,
        "finished_at": datetime.now().isoformat(),
        "tables": [],
        "errors": all_errors,
        "status": status,
        "record_count": record_count,
        "error_count": len(all_errors),
    }
